library/lib.ml

```diff
@@ -447,3 +371,247 @@ let discharge_proj_repr p =
+    let fs = Summary.freeze_staged_summaries stage ~marshallable:false in
+    add_entry (OpenedSection (prefix, fs))
+
+  let pop_path_prefix () = ()
+  let recalc_path_prefix () = ()
+
+end
+
+let add_discharged_leaf obj =
+  let newobj = Libobject.rebuild_object obj in
+  Libobject.cache_object (prefix(),newobj);
+  match Libobject.object_stage newobj with
+  | Summary.Stage.Synterp ->
+    SynterpActions.add_leaf_entry (AtomicObject newobj)
+  | Summary.Stage.Interp ->
+    InterpActions.add_leaf_entry (AtomicObject newobj)
+
+let add_leaf obj =
+  Libobject.cache_object (prefix(),obj);
+  match Libobject.object_stage obj with
+  | Summary.Stage.Synterp ->
+    SynterpActions.add_leaf_entry (AtomicObject obj)
+  | Summary.Stage.Interp ->
+    InterpActions.add_leaf_entry (AtomicObject obj)
+
+(** The [StagedLib] abstraction factors out the code dealing with Lib structure
+    that is common to all stages. *)
+module StagedLib(Actions : LibActions) = struct
+
+let add_entry node = Actions.add_entry node
+let add_leaf_entry obj = Actions.add_leaf_entry obj
+
+let open_section id = Actions.open_section id
+
+let find_opening_node id =
+  let entry = match Actions.get_lib_stk () with
+    | [] -> assert false
+    | (CompilingLibrary _, _) :: _ ->
+        CErrors.user_err Pp.(str "There is nothing to end.")
+    | (entry, _) :: _ -> entry
+  in
+  let id' = prefix_id (node_prefix entry) in
+  if not (Names.Id.equal id id') then
+    CErrors.user_err Pp.(str "Last block to end has name "
+      ++ Id.print id' ++ str ".");
+  entry
+
+let start_module = Actions.start_mod ~is_type:false
+let start_modtype = Actions.start_mod ~is_type:true None
+
+let end_mod ~is_type =
+  let (after,mark,before) = split_lib_at_opening (Actions.get_lib_stk ()) in
+  (* The errors here should not happen because we checked in the upper layers *)
+  let prefix, fs = match mark with
+    | OpenedModule (ty,_,prefix,fs) ->
+      if ty == is_type then prefix, fs
+      else error_still_opened (module_kind ty) prefix
+    | OpenedSection (prefix, _) -> error_still_opened "section" prefix
+    | CompilingLibrary _ -> CErrors.user_err (Pp.str "No opened modules.")
+  in
+  Actions.set_lib_stk before;
+  Actions.recalc_path_prefix ();
+  let after = classify_segment after in
+  (prefix, fs, after)
+
+let end_module () = end_mod ~is_type:false
+let end_modtype () = end_mod ~is_type:true
+
+(* Restore lib_stk and summaries as before the section opening, and
+   add a ClosedSection object. *)
+let close_section () =
+  let (secdecls,mark,before) = split_lib_at_opening (Actions.get_lib_stk ()) in
+  let fs = match mark with
+    | OpenedSection (_,fs) -> fs
+    | _ -> CErrors.user_err Pp.(str "No opened section.")
+  in
+  Actions.set_lib_stk before;
+  Actions.pop_path_prefix ();
+  let newdecls = List.map discharge_item secdecls in
+  Actions.close_section fs;
+  List.iter (Option.iter add_discharged_leaf) newdecls
+
+end
+
+module type StagedLibS = sig
```

Comment:
If you move this before the functor and use it as the functor's type we get more errors in merlin while developping instead of having to run ocamlc

Reply:
Done, thanks for the suggestion.